FEATURE: base feign-vertx on Vertx WebClient instead of bare vertx

hosuaby opened the pull request:
I made a number of changes to feign-vertx module:

-  Switch from bare vertx to high level WebClient. It will give the user more configuration options, including the possibility of enabling OAuth2/OIDC authentication (cause feign-vertx and feign-oauth2 are not and cannot be compatible).
- More reliable tests.
- Ensured compatibility with Vertx 5.

velo commented on vertx/feign-vertx/src/main/java/feign/VertxFeign.java:
```diff
@@ -122,13 +120,13 @@ public Builder invocationHandlerFactory(
+     * @param webClient vertx WebClient
      * @return this builder
      */
-    public Builder vertx(final Vertx vertx) {
```
My main concern about this change is breaking compatibility with previous version.


Would it be possible to keep this (probably as deprecate) and run `WebClient.create(vertx)`

Then we can instruct to use the new method and mark for removal on next major release

hosuaby replied:
No need for that. Because we never released the current version. The previous is still available, by the way, in the separate project that was merged into the main project recently.

velo commented on vertx/feign-vertx/src/main/java/feign/VertxFeign.java:
```diff
@@ -263,34 +261,6 @@ public Builder errorDecoder(final ErrorDecoder errorDecoder) {
     }
 
-    /**
-     * Sets request options using Vert.x {@link HttpClientOptions}.
-     *
-     * @param options {@code HttpClientOptions} for full customization of the underlying Vert.x
-     *     {@link HttpClient}
-     * @return this builder
-     */
-    public Builder options(final HttpClientOptions options) {
-      this.options = checkNotNull(options, "Argument options must be not null");
-      return this;
-    }
-
-    /**
-     * Sets request options using Feign {@link Request.Options}.
-     *
-     * @param options Feign {@code Request.Options} object
-     * @return this builder
-     */
-    @Override
-    public Builder options(final Request.Options options) {
-      checkNotNull(options, "Argument options must be not null");
-      this.options =
-          new HttpClientOptions()
-              .setConnectTimeout(options.connectTimeoutMillis())
-              .setIdleTimeout(options.readTimeoutMillis());
-      return this;
-    }
```
Same as above, is it possible to keep as deprecated!?

hosuaby replied:
Idem. Current version from core project was never released.

velo commented on vertx/feign-vertx/src/main/java/feign/VertxFeign.java:
```diff
@@ -307,22 +277,22 @@ public Builder timeout(long timeout) {
      * @param requestPreProcessor operation to execute on each request
      * @return updated request
      */
-    public Builder requestPreProcessor(UnaryOperator<HttpClientRequest> requestPreProcessor) {
```
the ask here would be the same, not sure if viable at all

hosuaby replied:
Idem. Current version from core project was never released.

velo replied:
Ow, sweet, perfect time to change then =)

velo commented on vertx/feign-vertx/src/main/java/feign/vertx/VertxHttpClient.java:
```diff
@@ -43,28 +44,25 @@
    */
   public VertxHttpClient(
-      final Vertx vertx,
-      final HttpClientOptions options,
+      final WebClient webClient,
       final long timeout,
-      final UnaryOperator<HttpClientRequest> requestPreProcessor) {
```
And lastly here too, keep old constructor, deprecate it, add new one

hosuaby replied:
Idem. Current version from core project was never released.

velo commented on vertx/pom.xml:
```diff
@@ -24,16 +24,22 @@
-    <jackson.version>2.18.2</jackson.version>
-    <slf4j-log4j12.version>2.0.0-alpha6</slf4j-log4j12.version>
-    <wiremock.version>2.35.2</wiremock.version>
+    <main.java.version>11</main.java.version>
```
is this needed?!

11 is the min version already

hosuaby replied:
Will check for it. Thanks

velo commented on vertx/pom.xml:
```diff
@@ -24,16 +24,22 @@
-  <description>Implementation of Feign on Vertx web client.</description>
+  <modules>
+    <module>feign-vertx</module>
+    <module>tests-vertx4</module>
+    <module>tests-vertx-common</module>
+    <module>tests-vertx5</module>
```
nice effort on test coverage 🥇

velo replied:
I think we can just call them `feign-vertx*-test`

hosuaby replied:
Thanks for the suggestion. I will study them carefully and see how I can improve tests. The most important is to avoid to produce the jar and run tests at same build. By doing so we will not be able to detect runtime binary uncompatibilities with different versions of Vertx (that arise quite offten).

hosuaby replied:
> nice effort on test coverage 🥇

Thanks! feign-vertx is a 10-year-old project. It had plenty of time to be tested in productions and covered with tests for the most of frequent problems.

hosuaby replied:
> I think we can just call them `feign-vertx*-test`

done

velo commented on vertx/tests-vertx5/src/test/java/feign/vertx/RetryingTest.java:
```diff
@@ -0,0 +1,141 @@
+import org.junit.jupiter.api.Test;
+
+@DisplayName("When server ask client to retry")
+class RetryingTest extends AbstractFeignVertxTest {
```
These look to be exactly the same

https://github.com/OpenFeign/feign/pull/2756/files#diff-618b10051163eb35607cd662bc192620411ccce4da0be7a99ee6dfb0b0ee4ad9L50

Why not move to test commons and just have a simplified version here?

velo reviewed (changes requested):
I think for tests you could keep them all on feign-vertx, generate the test-jar there, and on feign-vertx5-test you just extend the same tests under feign.vertx5 package....

Then you can make the module a top level project and if eventually vext breaks compatibility we can have multiples like we do with jax-rs

hosuaby commented:
> I think for tests you could keep them all on feign-vertx, generate the test-jar there, and on feign-vertx5-test you just extend the same tests under feign.vertx5 package....
> 
> Then you can make the module a top level project and if eventually vext breaks compatibility we can have multiples like we do with jax-rs

So here for the tests. What I managed and what not to achieve.

Achieved:

1) Moved all tests into the main module with application code feign-vertx.
2) Got rid of the module tests-vertx-common.
3) The main module feign-vertx now produces a test-jar to be imported in test modules feign-vertx4-test and feign-vertx5-test.
4) The module feign-vertx5-test contains no code. It will only be used to test binary compatibility with different versions of Vert.x 5.x.
5) The module feign-vertx4-test contains tests rewritten using the API of Vert.x 4.x. It is used to test binary compatibility with different versions of Vert.x 4.x.

What I could not achieve:

1) You suggested getting rid of the parent module, moving all the production code and tests to the root folder vertx, and producing two jars from it: one with production code and another with tests. By default, this is not possible with Maven and requires a ton of plugins to achieve. Due to architectural complexity, I did not pursue it.
2) I moved common test code used by both modules feign-vertx4-test and feign-vertx5-test into the test jar produced by feign-vertx, but I could not factorize it further. Vert.x 4.x and 5.x have different APIs, so I need to write different code to create a WebClient with the required configuration. I just decided to copy test classes, even if they are similar. Otherwise, it would require a complex test architecture, making them completely unreadable.

@velo  Please tell me what you think and what you would like to improve.
